lua/tardis/cl_options.lua

```diff
@@ -96,70 +96,126 @@ function TARDIS:ChangeOption(id,data)
+            elem:SetDecimals(0)
+        end
+        elem:SetValue(setting)
+        elem:GetTextArea():SetText(tostring(setting))
+        elem.OnValueChanged = function(self, val)
+            self.lastchange = CurTime()
+            self.lastchange_val = val
+        end
+        elem.Think = function(self)
+            if self.lastchange_val and CurTime() - self.lastchange > 0.1 then
+                TARDIS:SetSetting(id, self.lastchange_val)
+                self.lastchange_val = nil
             end
-            return buttons
-        end)
-        sections[k]=f
-    end
-    
-    local mainf=vgui.Create("DPanel",frame)
-    mainf:SetSize(frame:GetSize())
-    
-    self:LoadButtons(screen,mainf,function(frame)
-        local buttons={}
-        for k,v in ipairs(sections) do
-            local button = vgui.Create("DButton",frame)
-            button:SetText(v.name)
-            button:SetFont(TARDIS:GetScreenFont(screen, "Default"))
-            button.DoClick = function()
-                self:PushScreen(v.name,screen,frame,v)
+            if self.lastchange and CurTime() - self.lastchange > 0.3 then
+                self.lastchange = nil
+                self:SetValue(TARDIS:GetSetting(id))
+                self:GetTextArea():SetText(tostring(TARDIS:GetSetting(id)))
             end
-            table.insert(buttons,button)
         end
-        for k,v in ipairs(options) do
-            local name,id,data=v[1],v[2],v[3]
-            if not data.section then
-                local button = vgui.Create("DButton",frame)
-                button:SetText(name or id)
-                button:SetFont(TARDIS:GetScreenFont(screen, "Default"))
-                button.DoClick = function()
-                    self:ChangeOption(id,data)
-                end
-                table.insert(buttons,button)
+
+    elseif data.type=="color" then
+        elem = vgui.Create("DForm")
+        elem:SetLabel(text)
+        elem:SetExpanded(true)
+
+        local mixer = vgui.Create("DColorMixer")
+        mixer:SetText(text)
+        mixer:SetPalette(false)
+        mixer:SetAlphaBar(false)
+        mixer:SetWangs(true)
+        mixer:SetColor(setting)
+        mixer.ValueChanged = function(self, val)
+            TARDIS:SetSetting(id, val)
+            self.lastchange = CurTime()
+        end
+        elem.Think = function(self)
+            if self.lastchange and CurTime() - self.lastchange > 0.2 then
+                self.lastchange = nil
+                self:SetColor(TARDIS:GetSetting(id))
             end
         end
-        return buttons
-    end)
-end)
\ No newline at end of file
+
+        elem:AddItem(mixer)
+
+        local spacer = vgui.Create("DPanel")
+        spacer:SetTall(2)
+        elem:AddItem(spacer)
+
+    elseif data.type=="list" then
+        elem = vgui.Create("DLabel")
+
+        elem2 = vgui.Create("DComboBox")
+
+        if data.get_values_func ~= nil then
+            for k,v in pairs(data.get_values_func()) do
```

Comment:
Can this not be a direct k,v mapping here to optimise lookup + performance? Right now we seem to only be using the values and ignoring the key

Reply:
https://wiki.facepunch.com/gmod/DComboBox
There doesn't seem to be an `AddChoices()` function (or similar)
So we have to use AddChoice for each choice separately and have that extra loop there